---
title: "Introduction to m6ASeqTools"
author: "Hanna Lee"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to m6ASeqTools}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,       # show R code
  message = FALSE,   # suppress messages
  warning = FALSE,   # suppress warnings
  progress = FALSE   # hide knit progress lines
)
```

# Introduction
`m6ASeqTools` is an R package designed to streamline the post-processing and interpretation of site-level m6A predictions from m6Anet. It provides descriptive summaries of m6A distribution across transcripts, genes, biotypes, and transcript regions, and enables condition comparisons through a calculated weighted modification ratio. By integrating differential gene expression data, the package links methylation changes with expression differences, providing biotype-specific and region-specific insights into how m6A localization patterns relate to transcriptional regulation.

# Table of Contents


# Installation
`m6ASeqTools` can be installed on **Windows, macOS, and Linux**. It requires R (≥4.0).

```r
# Install devtools if needed
install.packages("devtools")

# Install Bioconductor dependencies
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(c(
    "AnnotationDbi",
    "org.Hs.eg.db",
    "GenomicFeatures",
    "GenomicRanges",
    "IRanges",
    "S4Vectors"
))

# Install m6ASeqTools from GitHub
devtools::install_github("hannalee809/m6ASeqTools")
```

# Input Data

The first step in using **m6ASeqTools** is importing the `data.site_proba.csv` file, which represents the final output from **m6Anet**.

> **Note:** Some columns in the raw file contain multiple values combined with a `|` separator. These must be split before using the data with this package.

For this tutorial, we provide two example datasets:

1. 100 predicted m6A sites from the **Gli36** glioma cell line.  
2. 100 predicted m6A sites from the same cell line following **IGF2BP2 siRNA knockdown**.

These example datasets allow you to explore the package workflow without needing your own experimental data.

```{r import-data}
library(m6ASeqTools)

# Load the example dataset included in the package
data("example_m6A_naive_gli36_cells")

# Preview the first few rows
head(example_m6A_naive_gli36_cells)
```

# Part 1: Descriptive Statistics

## Summary Report of m6Anet Output

After importing the `data.site_proba.csv` file in the required format, you can generate a summary report using the `summarize_m6anet_output()` function. This function produces an HTML report that includes the following:

1. **Overview table:** Total number of detected m6A sites, transcripts, and genes, as well as the number of sites, transcripts, and genes with a **modification probability > 90%**. These high-confidence sites are filtered and used for the rest of the report, and they should also be used for downstream analyses.
2. **High-confidence sites per transcript:** Table and histogram showing the distribution of the number of high-confidence sites per transcript.
3. **High-confidence transcripts per gene:** Table and histogram showing the distribution of high-confidence transcripts per gene.
4. **Transcript biotype distribution:** Table and pie chart(s) showing counts and frequencies of high-confidence transcripts by biotype.
5. **Modified k-mer distribution:** Table and bar graph showing counts and frequencies of modified k-mers.
6. **Transcript length distribution:** Table and density graph showing the lengths of modified transcripts, grouped into bins: 101–500 bp, 501–1000 bp, 1001–5000 bp, 5001–10000 bp, and >10000 bp.

```{r summarize-m6anet-output, eval=FALSE}
# Generate the summary HTML report
summarize_m6anet_output(
  example_m6A_naive_gli36_cells, 
  output_file = "naive_summary_report.html"
)
```

```{r summary-plot, echo=FALSE, out.width="100%", out.height="700px"}
knitr::include_graphics("naive_html_summary.png")
```

## Filter m6A Sites
m6Anet recommends retaining sites with probability_modified > 0.9.
This function applies a user-defined threshold (default = 0.9) and returns the filtered dataset.

> **Note:** Use this filtered dataset for downstream analyses.

```{r filter-m6a, eval = FALSE}
example_m6A_naive_gli36_cells <- filter_m6a_sites(example_m6A_naive_gli36_cells,
                                       probablity_modified=0.9)
```

## Transcript Region Analysis
These functions allow you to:

1. **Assign each m6A site to a transcript region**  
   (5'UTR, CDS, or 3'UTR)

2. **Compute its relative position within that region**, scaled 0–1.
This

3. **Visualize the global distribution of m6A** across these regions using a kernel density plot.

This workflow requires a GTF annotation file.

```{r tx-regions, eval = FALSE}
# Path to GTF annotation file
gtf_path <- "/path/to/gencode.v45.annotation.gtf"

# Step 1: Extract transcript region lengths (5'UTR, CDS, 3'UTR)
tx_df <- get_transcript_region_lengths(gtf_path)

# Step 2: Annotate each m6A site with its region and relative position
example_m6A_naive_gli36_cells <-
  map_relative_tx_regions_to_m6A(
    example_m6A_naive_gli36_cells,
    tx_df
  )

# Step 3: Visualize distribution of m6A across transcript regions
p <- plot_relative_positions(example_m6A_naive_gli36_cells, label = "naive") # add plot label if desired
```

```{r tx-plot, echo=FALSE, out.width="80%", out.height="400px"}
knitr::include_graphics("Transcript_Region_Density_naive.png")
```

## Chromosomal Annotation and Distribution
This function assigns each m6A site to its chromosome using human gene annotations from org.Hs.eg.db. Provide the m6A dataframe and optionally specify whether the results should be saved as a CSV and whether a chromosomal distribution plot should be generated.

```{r chr-dist, eval=FALSE}
calculate_chromosome_location(example_m6A_naive_gli36_cells,
                              output_csv=NULL, 
                              output_plot="naive_chr_distribution.png")
```

```{r chr-plot, echo=FALSE, out.width="80%", out.height="400px"}
knitr::include_graphics("naive_chr_distribution.png")
```

## Weighted Modification Ratio Calculation:
This metric summarizes transcript-level methylation by combining the number of modified sites, their modification ratios, and transcript length. It requires the annotated output from map_relative_tx_regions_to_m6A(), since transcript regions are used in downstream comparisons.

The **weighted modification ratio** per transcript is defined as:

$$
\text{Weighted Modification Ratio} = \frac{\sum_{i=1}^{n} r_i}{L}
$$

where:  

- \(r_i\) = modification ratio at site \(i\)  
- \(n\) = number of m6A sites in the transcript  
- \(L\) = length of the transcript (in nucleotides)

The function returns a dataframe containing one row per transcript with:

- total modification ratio across all sites
- counts of sites in 5′UTR, CDS, and 3′UTR
- the computed weighted modification ratio

```{r calc-wmr, eval=FALSE}
calculate_weighted_mod_ratio(example_m6A_naive_gli36_cells, mod_ratio_column = "mod_ratio")
```

```{r wmr-table, echo=FALSE}
library(kableExtra)

knitr::kable(
  data.frame(
    ensembl_transcript_name = c("ENST0001", "ENST0002"),
    ensembl_transcript_id = c("TXNAME1", "TXNAME2"),
    ensembl_gene_id = c("ENSG0001", "ENSG0002"),
    ensembl_gene_name = c("GENE1", "GENE2"),
    gene_biotype = c("protein_coding", "retained_intron"),
    transcript_length = c("2417", "2689"),
    sum_mod = c("1.81", "1.79"),
    n_sites = c("3", "3"),
    sites_in_5utr = c("0", "0"),
    sites_in_cds = c("0", "0"),
    sites_in_3utr = c("3", "3"),
    weighted_mod_ratio = c("0.00075", "0.00067")
  ),
  align = "l"
) %>%
  kable_styling(
    font_size = 10,
    full_width = FALSE
  )
```

# Part 2: Incorporating External Annotations

# Part 3: Comparative Analysis between Conditions

# Extra: m6A_seq_part1, m6A_seq_part2






